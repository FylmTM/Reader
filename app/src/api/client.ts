type Response<R> =
  | { status: "ok"; payload: R }
  | { status: "error"; message: string };

function post<T, R>(path: string, data: T): Promise<R> {
  return fetch(path, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

function get<R>(path: string): Promise<R> {
  return fetch(path, { method: "GET" })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

export default {
  get,
  post,
};
