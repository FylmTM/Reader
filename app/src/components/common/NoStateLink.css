span.link:hover {
    text-decoration: underline;
}
