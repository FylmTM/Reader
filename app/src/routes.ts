export const ALL_ROUTE = "/";

export const CATEGORY_ROUTE = "/category/:categoryId";
export const CATEGORY_POST_ROUTE = "/category/:categoryId/post/:postId";
export const FEED_ROUTE = "/feed/:feedId";
export const FEED_POST_ROUTE = "/feed/:feedId/post/:postId";
