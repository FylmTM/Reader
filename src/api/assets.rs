use std::borrow::Cow;
use std::ffi::OsStr;
use std::io::Cursor;
use std::path::{Path, PathBuf};

use rocket::http::ContentType;
use rocket::request::FromRequest;
use rocket::response;
use rocket::{Outcome, Request};

use crate::error::Error as E;

#[derive(RustEmbed)]
#[folder = "static"]
struct Assets;

const INDEX: &str = "index.html";

pub struct Asset {
    content: Cow<'static, [u8]>,
    content_type: Option<ContentType>,
}

impl<'a, 'r> FromRequest<'a, 'r> for Asset {
    type Error = E;

    fn from_request(request: &'a Request<'r>) -> rocket::request::Outcome<Self, E> {
        let path_raw = request.uri().path();

        // Do not try to serve assets for API.
        if path_raw.starts_with("/api") {
            return Outcome::Forward(());
        }

        let path = if path_raw == "/" {
            INDEX
        } else {
            if Path::new(path_raw).extension().is_none() {
                // Fallback all URL's that does not possibly represent file to index.html.
                // Required to support HTML5 history API for SPA application.
                INDEX
            } else {
                // We need to remove first '/' from url path, because assets paths are relative.
                &path_raw[1..]
            }
        };

        let file: Option<Cow<'static, [u8]>> = Assets::get(&path);
        let file_path = Path::new(path);
        let extension = file_path
            .extension()
            .and_then(OsStr::to_str)
            .expect("Asset extension must be present.");
        let content_type: Option<ContentType> = ContentType::from_extension(extension);

        file.map_or_else(
            || {
                debug!("Asset does not exist, forward request path={}.", path);
                Outcome::Forward(())
            },
            |content| {
                debug!("Serve asset path={}.", path);
                Outcome::Success(Asset {
                    content,
                    content_type,
                })
            },
        )
    }
}

#[get("/", rank = 98)]
pub fn index<'r>(asset: Asset) -> response::Result<'r> {
    get_asset(asset)
}

#[get("/<_file..>", rank = 99)]
pub fn assets<'r>(asset: Asset, _file: PathBuf) -> response::Result<'r> {
    get_asset(asset)
}

fn get_asset<'r>(asset: Asset) -> response::Result<'r> {
    match asset.content_type {
        Some(content_type) => response::Response::build()
            .header(content_type)
            .sized_body(Cursor::new(asset.content))
            .ok(),
        None => response::Response::build()
            .sized_body(Cursor::new(asset.content))
            .ok(),
    }
}
