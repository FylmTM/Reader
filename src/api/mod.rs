use crate::error::Error;
use rocket_contrib::json::Json;

pub mod authentication;
pub mod catchers;
pub mod endpoints;

#[derive(Serialize, Deserialize)]
pub struct ApiResponse<T> {
    status: String,
    payload: T,
}

impl<T> ApiResponse<T> {
    fn with(payload: T) -> ApiResponse<T> {
        ApiResponse {
            status: "ok".to_string(),
            payload,
        }
    }
}

type Response<T> = Result<Json<ApiResponse<T>>, Error>;

fn ok<T>(data: T) -> Response<T> {
    Ok(Json(ApiResponse::with(data)))
}
