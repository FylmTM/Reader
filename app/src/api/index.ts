import {
  CategoriesWithFeeds,
  PostsUnreadCount,
  Page,
  Post,
  PostContent,
  PostsSection,
  User,
} from "../domain";
import client from "./client";

function getCurrentUser(): Promise<User> {
  return client.get("/api/v1/auth/current_user");
}

function login(apiKey: string): Promise<User> {
  return client.post("/api/v1/auth/login", { api_key: apiKey });
}

function logout(): Promise<void> {
  return client.post("/api/v1/auth/logout", {});
}

function getCategoriesWithFeeds(): Promise<CategoriesWithFeeds> {
  return client.get("/api/v1/categories_with_feeds");
}

function getCategoriesWithFeedsUnreadCounts(): Promise<PostsUnreadCount> {
  return client.get("/api/v1/posts/unread_count");
}

function getPosts(
  section: PostsSection,
  isUnreadOnly: boolean,
  fromPostId?: number,
): Promise<Page<Post>> {
  const query = [];

  if (fromPostId != null) {
    query.push(`from_post_id=${fromPostId}`);
  }
  if (isUnreadOnly && section.type !== "read-later") {
    query.push(`is_read=false`);
  }

  switch (section.type) {
    case "read-later":
      query.push(`is_read_later=true`);
      break;
    case "all":
      // nothing
      break;
    case "category":
      query.push(`category_id=${section.categoryId}`);
      break;
    case "feed":
      query.push(`feed_id=${section.feedId}`);
      break;
  }

  const queryString = query.length === 0 ? "" : `?${query.join("&")}`;
  return client.get(`/api/v1/posts${queryString}`);
}

function postsMarkAsRead(
  section: PostsSection,
  fromPostId: number,
): Promise<Page<Post>> {
  const query = [];
  query.push(`from_post_id=${fromPostId}`);

  switch (section.type) {
    case "read-later":
      throw new Error("Mark all as read not supported for read-later");
    case "all":
      // nothing
      break;
    case "category":
      query.push(`category_id=${section.categoryId}`);
      break;
    case "feed":
      query.push(`feed_id=${section.feedId}`);
      break;
  }

  const queryString = query.length === 0 ? "" : `?${query.join("&")}`;
  return client.get(`/api/v1/posts/mark_as_read${queryString}`);
}

function getPostContent(postId: number): Promise<PostContent> {
  return client.get(`/api/v1/posts/${postId}/content`);
}

function markAsRead(postId: number, isRead: Boolean): Promise<void> {
  return client.post(`/api/v1/posts/${postId}/is_read`, {
    is_read: isRead,
  });
}

function markAsReadLater(postId: number, isRead: Boolean): Promise<void> {
  return client.post(`/api/v1/posts/${postId}/is_read_later`, {
    is_read_later: isRead,
  });
}

export default {
  getCurrentUser,
  login,
  logout,
  getCategoriesWithFeeds,
  getCategoriesWithFeedsUnreadCounts,
  getPosts,
  postsMarkAsRead,
  getPostContent,
  markAsRead,
  markAsReadLater,
};
