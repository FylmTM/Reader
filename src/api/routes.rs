use std::ops::Add;

use rocket::http::{Cookie, Cookies};
use rocket_contrib::json::Json;

use crate::api::{ok, Response};
use crate::db::{self, Queries};

#[derive(Serialize, Deserialize)]
pub struct ApiKey {
    pub api_key: String,
}

#[post("/api/v1/auth/login", data = "<api_key>")]
pub fn auth_login(
    api_key: Json<ApiKey>,
    conn: db::PoolConnection,
    mut cookies: Cookies,
) -> Response<db::User> {
    let api_key = api_key.0.api_key;
    let user = conn.find_user_by_api_key(&api_key)?;
    cookies.add(
        Cookie::build("api_key", api_key)
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::days(365)))
            .finish(),
    );

    ok(user)
}

#[post("/api/v1/auth/logout")]
pub fn auth_logout(mut cookies: Cookies) -> Response<()> {
    cookies.add(
        Cookie::build("api_key", "")
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::milliseconds(1)))
            .finish(),
    );

    ok(())
}

#[get("/api/v1/auth/current_user")]
pub fn auth_current_user(user: db::User) -> Response<db::User> {
    ok(user)
}

#[get("/api/v1/feeds/update")]
pub fn feeds_update(_user: db::User, mut conn: db::PoolConnection) -> Response<()> {
    crate::feeds::update(&mut conn);
    ok(())
}

#[get("/api/v1/categories_with_feeds")]
pub fn categories_with_feeds(
    user: db::User,
    conn: db::PoolConnection,
) -> Response<db::CategoriesWithFeeds> {
    let categories = conn.find_categories_by_user(user.id)?;

    let mut categories_with_feeds = Vec::with_capacity(categories.len());
    for category in categories {
        let feeds = conn.find_feeds_by_category(category.id)?;
        categories_with_feeds.push(db::CategoryWithFeeds { category, feeds })
    }

    ok(categories_with_feeds)
}

#[get("/api/v1/posts?<from_post_id>&<is_read_later>&<is_read>&<category_id>&<feed_id>")]
pub fn posts(
    from_post_id: Option<i64>,
    is_read_later: Option<bool>,
    is_read: Option<bool>,
    category_id: Option<i64>,
    feed_id: Option<i64>,
    user: db::User,
    conn: db::PoolConnection,
) -> Response<db::Page<db::UserPost>> {
    let posts = conn.find_posts(
        user.id,
        from_post_id,
        is_read_later,
        is_read,
        category_id,
        feed_id,
    )?;
    ok(posts)
}

#[get("/api/v1/posts/unread_count")]
pub fn posts_unread_count(
    user: db::User,
    conn: db::PoolConnection,
) -> Response<db::UserPostsUnreadCount> {
    let count = conn.count_posts_unread(user.id)?;
    ok(count)
}

#[get("/api/v1/posts/mark_as_read?<from_post_id>&<category_id>&<feed_id>")]
pub fn posts_mark_as_read(
    from_post_id: i64,
    category_id: Option<i64>,
    feed_id: Option<i64>,
    user: db::User,
    conn: db::PoolConnection,
) -> Response<()> {
    conn.posts_mark_as_read(user.id, from_post_id, category_id, feed_id)?;
    ok(())
}

#[get("/api/v1/posts/<post_id>/content")]
pub fn post_content(
    post_id: i64,
    user: db::User,
    conn: db::PoolConnection,
) -> Response<db::UserPostContent> {
    let content = conn.find_post_content_by_user(user.id, post_id)?;
    ok(content)
}

#[derive(Serialize, Deserialize)]
pub struct PostIsReadData {
    pub is_read: bool,
}

#[post("/api/v1/posts/<post_id>/is_read", data = "<data>")]
pub fn post_is_read(
    post_id: i64,
    data: Json<PostIsReadData>,
    user: db::User,
    conn: db::PoolConnection,
) -> Response<()> {
    conn.set_post_is_read(user.id, post_id, data.0.is_read)?;
    return ok(());
}

#[derive(Serialize, Deserialize)]
pub struct PostIsReadLaterData {
    pub is_read_later: bool,
}

#[post("/api/v1/posts/<post_id>/is_read_later", data = "<data>")]
pub fn post_is_read_later(
    post_id: i64,
    data: Json<PostIsReadLaterData>,
    user: db::User,
    conn: db::PoolConnection,
) -> Response<()> {
    conn.set_post_is_read_later(user.id, post_id, data.0.is_read_later)?;
    return ok(());
}
