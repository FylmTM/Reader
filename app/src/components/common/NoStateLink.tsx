import React, { FC, useCallback } from "react";
import "./NoStateLink.css";

type NativeAnchorElement = React.DetailedHTMLProps<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  HTMLAnchorElement
> & {
  onClickHref?: string;
};

export function navigate(href: string | undefined) {
  if (href != null) {
    // eslint-disable-next-line no-restricted-globals
    history.pushState({}, "", href);
  }
}

export const NoStateLink: FC<NativeAnchorElement> = function NoStateLink({
  href,
  onClickHref,
  onClick,
  children,
  ...props
}) {
  const handleClick = useCallback(
    event => {
      // ignores the navigation when clicked using right mouse button or
      // by holding a special modifier key: ctrl, command, win, alt, shift
      if (
        event.ctrlKey ||
        event.metaKey ||
        event.altKey ||
        event.shiftKey ||
        event.button !== 0
      ) {
        return;
      }

      event.preventDefault();
      navigate(onClickHref || href);
      onClick && onClick(event);
    },
    [href, onClick, onClickHref]
  );

  return (
    <a href={href} onClick={handleClick} {...props}>
      {children}
    </a>
  );
};

export const NoStateSpanLink: FC<NativeAnchorElement> = function NoStateSpanLink({
  href,
  onClick,
  children,
  className,
  ...props
}) {
  const handleClick = useCallback(
    event => {
      // ignores the navigation when clicked using right mouse button or
      // by holding a special modifier key: ctrl, command, win, alt, shift
      if (
        event.ctrlKey ||
        event.metaKey ||
        event.altKey ||
        event.shiftKey ||
        event.button !== 0
      ) {
        return;
      }

      event.preventDefault();
      navigate(href);
      onClick && onClick(event);
    },
    [href, onClick]
  );

  return (
    <span onClick={handleClick} className={`link ${className}`} {...props}>
      {children}
    </span>
  );
};
