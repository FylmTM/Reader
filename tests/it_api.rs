#![feature(proc_macro_hygiene, decl_macro)]
#[macro_use]
extern crate rocket;

#[macro_use]
extern crate insta;

use common::client::{self, RequestOperations, ResponseOperations};
use reader::api::routes::*;

pub mod common;

#[test]
fn test_auth_login_failed() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "invalid".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity())
}

#[test]
fn test_auth_login_success() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "api_key".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity());
    assert_eq!(response.cookies().len(), 1);
}

#[test]
fn test_auth_logout_success() {
    let client = client::authenticated();

    let mut response = client.post("/api/v1/auth/logout").dispatch();
    assert_eq!(response.entity(), client::api_ok(()));
    assert_eq!(response.cookies().len(), 1);

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user_failed_if_not_authenticated() {
    let client = client::get();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_categories_with_feeds() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/categories_with_feeds").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_all() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_read_later() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?is_read_later=true").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_category() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?category_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_feed() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?feed_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_post_content() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts/1/content").dispatch();
    assert_json_snapshot!(response.json_entity());
}
