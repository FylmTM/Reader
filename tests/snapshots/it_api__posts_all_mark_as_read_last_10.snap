---
source: tests/it_api.rs
expression: response.json_entity()
---
[
  "200 OK",
  {
    "payload": {
      "all": 154,
      "categories": {
        "1": 74,
        "2": 80
      },
      "feeds": {
        "1": 20,
        "2": 4,
        "3": 50,
        "4": 20,
        "5": 10,
        "6": 10,
        "7": 10,
        "8": 20,
        "9": 10
      }
    },
    "status": "ok"
  }
]
